Add arbitrary hosts by docker label

1ed opened the pull request:
Hello!

First of all thank you for this great tool!
We need to attach arbitrary host names to some containers e.g. for using google oauth. This adds the ability to provide it via a docker label called: `com.costela.docker-etchosts.host`

This is my first go code ever.

I thinking about adding support for json arrays like something this:  

```go
package main

import (
	"encoding/json"
	"fmt"
)

func main() {
	hostsJson := `["s"]`
	var hosts []string
	err := json.Unmarshal([]byte(hostsJson), &hosts)
	if err != nil {
		fmt.Println(err)
	}
	for _, host := range hosts {
		fmt.Printf("%s", host)
	}
}
```

TODO:

- [x] json array?
- [x] tests?
- [x] readme update

1ed committed:
Add arbitrary hosts by docker label

1ed commented:
@costela could you have a look please?

costela commented:
Hey @1ed! Thanks for the contribution! Unfortunately I'm a bit swamped at the moment and will only have time to look at this next week :disappointed: 
I'll get back to you ASAP.

1ed commented:
Did you get a chance to look into this?

costela commented on README.md:
```diff
@@ -39,19 +39,23 @@ Entries are created for each container network with the following names:
 
 Each container will thereforr have up to 4 entries per alias: CONTAINER_ALIAS, CONTAINER_ALIAS.PROJECT, CONTAINER_ALIAS.NETWORK_NAME, CONTAINER_ALIAS.PROJECT.NETWORK_NAME
 
+Arbitrary hosts entries can be added via a custom label (`com.costela.docker-etchosts.add_hosts`) by specifying a single or array of host names.
```
nitpicking, but I think we could use `extra_hosts` here to make it closer to what docker-compose already uses? 
```suggestion
Arbitrary hosts entries can be added via a custom label (`net.costela.docker-etchosts.extra_hosts`) by specifying a single or array of host names.
```

costela commented on README.md:
```diff
@@ -39,19 +39,23 @@ Entries are created for each container network with the following names:
   someservice:
     ...
+    labels:
+        - 'com.costela.docker-etchosts.add_hosts=["a.example.com", "b.example.com"]'
```
(see above)
```suggestion
        - 'net.costela.docker-etchosts.extra_hosts=["a.example.com", "b.example.com"]'
```

costela commented on docker.go:
```diff
@@ -89,6 +90,28 @@ func getIPsToNames(client dockerClienter, id string) (ipsToNamesMap, error) {
 			names = appendNames(names, name)
 		}
 
+		if label, ok := containerFull.Config.Labels["com.costela.docker-etchosts.add_hosts"]; ok {
```
(see above)
```suggestion
		if label, ok := containerFull.Config.Labels["net.costela.docker-etchosts.extra_hosts"]; ok {
```
but maybe we could put this into a constant? At least the prefix?

costela commented on docker.go:
```diff
@@ -89,6 +90,28 @@ func getIPsToNames(client dockerClienter, id string) (ipsToNamesMap, error) {
 		}
 
+		if label, ok := containerFull.Config.Labels["com.costela.docker-etchosts.add_hosts"]; ok {
+			if (strings.HasPrefix(label, "[")) {
```
maybe we can be a bit more defensive here?
```suggestion
			label = strings.TrimSpaces(label)
			if (strings.HasPrefix(label, "[")) {
```

costela commented on docker.go:
```diff
@@ -89,6 +90,28 @@ func getIPsToNames(client dockerClienter, id string) (ipsToNamesMap, error) {
+					log.Errorf("error parsing JSON: %s", err)
+				}
+				names = append(names, parsed...)
+			} else if (strings.HasPrefix(label, "\"")) {
```
nitpicking :wink: :
```suggestion
			} else if (strings.HasPrefix(label, `"`)) {
```

costela commented on docker.go:
```diff
@@ -89,6 +90,28 @@ func getIPsToNames(client dockerClienter, id string) (ipsToNamesMap, error) {
+					log.Errorf("error parsing JSON: %s", err)
+				}
+				names = append(names, parsed)
+			} else if (strings.HasPrefix(label, "{")) {
+				log.Errorf("JSON objects are not supported: %s", label)
+			} else {
+				names = append(names, label)
+			}
```
Maybe we can also be a bit more defensive here?
We could check in the first condition against a valid domain regex (something optimistic like `[a-zA-Z][a-zA-Z0-9.-]*` - from hosts(5) -  will probably work for 90% of cases) and error for everything else?
If you think the regexp is too restrictive, we could at least filter out things that would break the `/etc/hosts` file, like newlines, just to be on the safe side.
WDYT?

1ed replied:
I made it a little bit stricter based on the manual. Thanks!

costela reviewed (changes requested):
hi @1ed!
Sorry it took so long to look at your PR! :sweat: 

I have a couple of comments. Maybe you can take a look and tell me what you think?

Thanks again for taking the time to improve this little project! :+1:

1ed committed:
Use extra_hosts instead of add_hosts

1ed committed:
Add hostname validation

1ed commented:
hi @costela! Thanks for your review! I addressed all your comments. And in the meantime I migrated the Travis CI build over to Github Actions and go releaser in my fork here https://github.com/1ed/docker-etchosts/blob/master/.github/workflows/go.yml. It creates automatic Github and Dockerhub releases after a new tag was pushed (see https://github.com/1ed/docker-etchosts/actions/runs/1087507407 and https://github.com/1ed/docker-etchosts/releases/tag/0.2.0). What do you think, should I do the same here?

costela commented:
LGTM! thanks again for the work and patience! :+1:

costela commented:
@1ed Oh, and as for GH-Actions: if you already did the work, then please, feel free to open a PR for that as well! :+1:

1ed commented:
Thanks! Here you go https://github.com/costela/docker-etchosts/pull/3
